upup/pkg/fi/task.go

```diff
@@ -50,6 +50,14 @@ type ModelBuilder interface {
 	Build(context *ModelBuilderContext) error
 }
 
+// HasDeletions is a ModelBuilder that creates tasks to delete cloud objects that no longer exist in the model.
+type HasDeletions interface {
+	ModelBuilder
+	// FindDeletions finds cloud objects that are owned by the cluster but no longer in the model and creates tasks to delete them.
+	// It is not called for the Terraform or Cloudformation targets.
+	FindDeletions(context *ModelBuilderContext, cloud Cloud) error
+}
+
```

Comment:
We already have this:

https://github.com/kubernetes/kops/blob/da6eee03b0cfb96fb3b5622a9d7fc907ebe927be/upup/pkg/fi/deletions.go#L19-L28

would it make sense to consolidate these into one interface? I'm not sure I understand the difference between how they're both used.

Reply:
`ProducesDeletions` finds everything of a type that is owned by the cluster and creates a `Deletion` that gets rid of it. It is only useful if you're deleting the cluster because it has no concept of objects owned by the cluster that need to be kept.

`HasDeletions` is a subclass of `ModelBuilder` that, during a second pass during apply_cluster, looks at the `Task`s representing the expected state. It finds objects owned by the cluster that have no corresponding expected `Task` and adds additional `Task`s to delete them.

Reply:
It might make sense to later consolidate `ProducesDeletions` into `HasDeletions`, but there are a lot of deletions that only happen during cluster teardown and the `Task` interface has more requirements than the `Deletion` interface. I see such consolidation as potential future work.